### PR TITLE
Vision parallel

### PR DESCRIPTION
**Pull request by Flova**
When I discovered the Issue #96 I tried to fix it by improving the thread-safety. This resulted in this pull request. It also fixes issue #90 .

### CODE REVIEW

File: bitbots_vision/src/bitbots_vision/vision.py [hunk 1]
```diff
@@ -86,17 +93,50 @@ def __init__(self):
...
-        rospy.spin()
+        # Run the vision main loop
+        self.main_loop()
+
+    def main_loop(self):
+        """
+        Main loop that processed the images and configuration changes
+        """
+        while not rospy.is_shutdown():
+            # Lookup if theres another configuration avalabile
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/vision.py]:
```suggestion
            # Lookup if there is another configuration available
```

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py [hunk 2]
```diff
@@ -138,8 +138,8 @@ def image_callback(self, image_msg):
 
         # Drops old images
         image_age = rospy.get_rostime() - image_msg.header.stamp
-        if 0.1 < image_age.to_sec() < 1000.0:
-            rospy.loginfo('Vision: Dropped incoming Image-message, because its too old!')
+        if 1.0 < image_age.to_sec() < 1000.0:
+            rospy.logwarn_throttle(2, 'Vision: Dropped incoming Image-message, because its too old! ({} sec)'.format(image_age.to_sec()))
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/dynamic_color_space.py]: Is this related to the PR?

File: bitbots_vision/src/bitbots_vision/vision.py [hunk 3]
```diff
@@ -86,17 +93,50 @@ def __init__(self):
...
+                self.configure_vision(*reconfigure_data)
+            # Check if a new image is avalabile
+            elif self.transfer_image_msg is not None:
+                # Copy image from shard memory
+                image_msg = deepcopy(self.transfer_image_msg)
+                self.transfer_image_msg = None
+                # Run the vision pipeline
+                self.handle_image(image_msg)
+                # Now the first image has been processed
+                self._first_image_callback = False
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision.py]: do we need a sleep here i the else part?

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py (hunk 2 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/dynamic_color_space.py]: No not really, I was wondering why this showed up here. It is from another pull request yet merged and closed pull request.

File: bitbots_vision/src/bitbots_vision/vision.py (hunk 3 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision.py]: You mean, that the loop should not spin at insanely high rates if there's nothing to? We cloud add it, although I didn't noticed any weird behavior in the testing.

File: bitbots_vision/src/bitbots_vision/vision.py [hunk 4]
```diff
@@ -86,17 +93,50 @@ def __init__(self):
...
+                # Copy config from shard memory
+                self.transfer_reconfigure_data_read_flag = True
+                reconfigure_data = deepcopy(self.transfer_reconfigure_data)
+                self.transfer_reconfigure_data_read_flag = False
+                self.transfer_reconfigure_data = None
+                # Run vision reconfiguration
+                self.configure_vision(*reconfigure_data)
+            # Check if a new image is avalabile
+            elif self.transfer_image_msg is not None:
+                # Copy image from shard memory
```
Comment (johagge) [inline bitbots_vision/src/bitbots_vision/vision.py]:
```suggestion
                # Copy image from shared memory
```
